### PR TITLE
Stop sending collectd metrics when buffer queue size exceeds limit

### PR DESCRIPTION
**Pull request by sugr4**
@Microsoft/omsagent-devs

### CODE REVIEW

File: source/code/plugins/collectd_lib.rb [hunk 1]
```diff
@@ -38,21 +38,27 @@ def transform(record, hostname)
...
-          "DataItems"=>data_items
-        }
-         return wrapper
-       else
-         # no data items, send a empty array that tells ODS
-         # output plugin to not the data
-         return {}
-       end
+      if record !=nil and record["error_class"]=="Fluent::BufferQueueLimitError"
+        $log.warn "Buffer Queue limit exceeded, collectD metrics not being sent to OMS"
```
Comment (niroyb) [inline source/code/plugins/collectd_lib.rb]: Consider OMSlog::warn_once to prevent spamming the log.

File: source/code/plugins/collectd_lib.rb [hunk 2]
```diff
@@ -38,21 +38,27 @@ def transform(record, hostname)
...
-          "IPName"=>"LogManagement",
-          "DataItems"=>data_items
-        }
-         return wrapper
-       else
-         # no data items, send a empty array that tells ODS
-         # output plugin to not the data
-         return {}
-       end
+      if record !=nil and record["error_class"]=="Fluent::BufferQueueLimitError"
```
Comment (niroyb) [inline source/code/plugins/collectd_lib.rb]: Is `error_class` part of the collectd record or the fluentd monitor one?

**Comment by niroyb** [comment]
Signed off

File: source/code/plugins/collectd_lib.rb (hunk 2 above)
Comment (sugr4) [inline source/code/plugins/collectd_lib.rb]: The error_class I used is part of the collectd record.

016-05-24 15:12:36 -0700 [warn]: dump an error event: error_class=NoMethodError error="undefined method `each_index' for nil:NilClass" tag="fluent.warn" time=1464127956 record={"error_class"=>"Fluent::BufferQueueLimitError", "error"=>"queue size exceeds limit", "tag"=>"oms.collectd", "message"=>"emit transaction failed: error_class=Fluent::BufferQueueLimitError error=\"queue size exceeds limit\" tag=\"oms.collectd\"", "Timestamp"=>"2016-05-24T22:12:36.278Z"}

**Pull request merged**